### PR TITLE
feat: add support for multiple root directories under a single prefix

### PR DESCRIPTION
**Pull request by nigelhanlon**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

This Pull Request aims to add support for multiple `root` directories under a single prefix as per feature suggestion #144. 

Since the `send` module used only accepts a single root path, I have implemented a recursive search approach using the stream error handler. The logic works as follows:

- We take the first path in the root path array and instantiate an instance of `send`.
- If we find the matching file on that path we return it.
- If we find a matching directory on that path we return a listing or index file.
- If we do not find a match we call `pumpSendToReply` again with the same parameters but increment the array offset.
- If we exhaust the root path array we return 404.
 
This approach allows us to return as soon as we have found a matching file/directory without considering all paths and also solves the problem of multiple files with the same name. The first match wins.

### CODE REVIEW

**Review by mcollina — COMMENTED** [review]
code looks good!

**Comment by nigelhanlon** [comment]
Ran some rough benchmarks to make sure these changes didn't cause any major regressions in performance. 

My machine is a 2017 MacBook Pro, 3.1 GHz Dual-Core Intel Core i5, 8GB RAM.

## Master branch for comparison

On master, my laptop manages roughly 3350 req/seq.  

```
autocannon -c 100 -d 40 -p 10 localhost:3000/index.html
Running 40s test @ http://localhost:3000/index.html
100 connections with 10 pipelining factor

Stat         Avg     Stdev  Max
Latency (ms) 29.6    87.26  942.82
Req/Sec      3348.63 681.4  4077
Bytes/Sec    1.71 MB 348 kB 2.08 MB

134k requests in 40s, 68.2 MB read
```

## Array of root paths, serving from first path

```
autocannon -c 100 -d 40 -p 10 localhost:3000/index.html
Running 40s test @ http://localhost:3000/index.html
100 connections with 10 pipelining factor

Stat         Avg     Stdev  Max
Latency (ms) 28.74   81.73  823.3
Req/Sec      3449.33 479.68 4118
Bytes/Sec    1.76 MB 238 kB 2.1 MB

138k requests in 40s, 70.2 MB read
```

## Array of root paths, serving from second path

Performance is slower since we take a time penalty when we don't find the file in the first path. 

This penalty should increase with every miss.

```
autocannon -c 100 -d 40 -p 10 localhost:3000/test.html
Running 40s test @ http://localhost:3000/test.html
100 connections with 10 pipelining factor

Stat         Avg     Stdev  Max
Latency (ms) 32.06   95.35  1232.6
Req/Sec      3098.78 681.15 3964
Bytes/Sec    1.34 MB 293 kB 1.71 MB

124k requests in 40s, 53.5 MB read
```

**Comment by mcollina** [comment]
is this still draft?

**Comment by nigelhanlon** [comment]
> is this still draft?

This should be ready for final review. Happy to make any changes as needed or suggested.

File: index.js [hunk 1]
```diff
@@ -163,17 +173,24 @@ function fastifyStatic (fastify, opts, next) {
       }
     } else {
       const globPattern = typeof opts.wildcard === 'string' ? opts.wildcard : '**/*'
+
+      // TODO: This won't work with an array of roots.
```
Comment (mcollina) [inline index.js:177]: can you resolve this TODO?

File: index.js [hunk 2]
```diff
@@ -163,17 +173,24 @@ function fastifyStatic (fastify, opts, next) {
...
+      }
+
       glob(path.join(sendOptions.root, globPattern), { nodir: true }, function (err, files) {
         if (err) {
           return next(err)
         }
         const indexDirs = new Set()
         const indexes = typeof opts.index === 'undefined' ? ['index.html'] : [].concat(opts.index || [])
         for (let file of files) {
+          // TODO: This won't work with an array of roots.
```
Comment (mcollina) [inline index.js:189]: can you resolve this TODO?

File: index.d.ts [hunk 3]
```diff
@@ -31,7 +31,7 @@ interface ListOptions {
 }
 
 export interface FastifyStaticOptions {
-  root: string;
+  root: string | string[];
```
Comment (mcollina) [inline index.d.ts:34]: we also test our types, please add a test for this type as well

**Review by mcollina — CHANGES_REQUESTED** [review]
Good work!

File: index.js [hunk 4]
```diff
@@ -163,17 +173,24 @@ function fastifyStatic (fastify, opts, next) {
...
           file = file.replace(sendOptions.root.replace(/\\/g, '/'), '').replace(/^\//, '')
-          const route = encodeURI(prefix + file).replace(/\/\//g, '/')
+          const route = (prefix + file).replace(/\/\//g, '/')
```
Comment (simoneb) [inline index.js:191]: is this change that removed encodeURI tested somehow? I'm not sure if this was necessary in the first place, but if we don't know for sure we should assume that it was necessary.

Comment (nigelhanlon) [reply index.js:191]: Well spotted. Not sure why that changed but it should indeed be `encodeURI`.

File: index.js [hunk 5]
```diff
@@ -108,6 +112,12 @@ function fastifyStatic (fastify, opts, next) {
           if (opts.list && dirList.handle(pathname, opts.list)) {
             return dirList.send({ reply, dir: dirList.path(opts.root, pathname), options: opts.list, route: pathname })
           }
+
+          // root paths left to try?
+          if (Array.isArray(rootPath) && rootPathOffset <= rootPath.length) {
```
Comment (simoneb) [inline index.js:117]: are you sure <= is right? if `rootPathOffset === rootPath.length` and we're calling `pumpSendToReply` with `rootPathOffset + 1`, aren't we exceeding the bounds of the `rootPath` array?

Comment (nigelhanlon) [reply index.js:117]: Yes, you are right. The offset is always going to be one smaller than the length of the array. 

I can change it to:

```
if (Array.isArray(rootPath) && rootPathOffset < (rootPath.length - 1))
```
The only unique case would be when there is only one element in the array and we get the offset = 0 and the length = 0 (length - 1). `0 < 0` evaluates to false but it seems a little messy.

Comment (simoneb) [reply index.js:117]: what about `offset < length`?

Comment (nigelhanlon) [reply index.js:117]: The problem is the condition check is at the end so we essentially have a do-while loop. `offset < length` will run the pump function on more time past the bounds of the array. If I move the increment before the check then it should work.

File: index.js [hunk 6]
```diff
@@ -163,39 +174,50 @@ function fastifyStatic (fastify, opts, next) {
...
+      }
+
+      // TODO: this causes fastify to fail when running with multiple root paths.
+      if (Array.isArray(sendOptions.root)) {
+        for (let i = 0; i < sendOptions.root.length; i++) {
+          globber(sendOptions.root[i])
+        }
+      } else {
+        globber(sendOptions.root)
+      }
```
Comment (nigelhanlon) [inline index.js:220]: This loop seems to cause issues when registering fastify routes for each of the matching glob paths. The first iteration is fine but the second fails.

File: index.d.ts (hunk 3 above)
Comment (nigelhanlon) [reply index.d.ts:34]: Added a small addition to the `test/types/index.ts` file to check the type works.

File: index.js (hunk 6 above)
Comment (nigelhanlon) [reply index.js]: We now use `async.eachSeries`.

File: index.js [hunk 7]
```diff
@@ -163,39 +174,48 @@ function fastifyStatic (fastify, opts, next) {
...
+            }
+          })
+          cb()
         })
-        next()
-      })
+      }
+
+      if (Array.isArray(sendOptions.root)) {
+        async.eachSeries(sendOptions.root, globber, next)
```
Comment (mcollina) [inline index.js:215]: I would actually prefer to use `util.promisify` to get a promise based version of `glob` and use async/await in a for loop instead.

Comment (nigelhanlon) [reply index.js:215]: No problem, I will update.

File: index.js [hunk 8]
```diff
@@ -163,52 +174,78 @@ function fastifyStatic (fastify, opts, next) {
       }
     } else {
       const globPattern = typeof opts.wildcard === 'string' ? opts.wildcard : '**/*'
-      glob(path.join(sendOptions.root, globPattern), { nodir: true }, function (err, files) {
-        if (err) {
-          return next(err)
-        }
+      const globPromise = util.promisify(glob)
```
Comment (mcollina) [inline index.js:177]: move this to the top of the file

**Review by mcollina — APPROVED** [review]
lgtm

File: index.js (hunk 8 above)
Comment (nigelhanlon) [reply index.js]: Done 👍

**Review by mcollina — APPROVED** [review]
lgtm

File: README.md [hunk 9]
```diff
@@ -59,6 +59,9 @@ The absolute path of the directory that contains the files to serve.
 The file to serve will be determined by combining `req.url` with the
 provided root directory.
 
+You can also provide an array of directories containing files to serve.
+This is useful for serving multiple static directories under a single prefix. Files are served in a "first found, first served manner", so the order in which you list the directories is important. For best performance, you should always list your main asset directory first. Duplicate paths will raise an error.
```
Comment (simoneb) [inline README.md:63]: nit (double quotes position): 

```suggestion
This is useful for serving multiple static directories under a single prefix. Files are served in a "first found, first served" manner, so the order in which you list the directories is important. For best performance, you should always list your main asset directory first. Duplicate paths will raise an error.
```

File: example/server-multi-dir-list.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+list
+
+<ul>
+{{#files}}
+<li><a href="{{href}}" target="_blank">{{name}}</a></li>
+{{/files}}
+</ul>
+</body></html>
+`
```
Comment (simoneb) [inline example/server-multi-dir-list.js:27]: formatting: can you please add some indentation to this so it looks more like a properly formatted html file? also, do we need to have it inline or would it make sense to have it as a file on the file system?

File: example/server-multi-dir-list.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+fastify
+  .register(require('..'), {
+    root: [path.join(__dirname, '/public'), path.join(__dirname, '/public2')],
+    prefixAvoidTrailingSlash: true,
+    wildcard: '**/*.css',
+    list: {
+      format: 'html',
+      names: ['index', 'index.html', 'index.htm', '/'],
+      render: (dirs, files) => handlebarTemplate({ dirs, files })
+    }
```
Comment (simoneb) [inline example/server-multi-dir-list.js:39]: can you maybe a explain with some comments why you're setting all these options? being it an example, it would benefit the readers in understanding what this is doing, as an extension to the docs basically

File: index.js [hunk 12]
```diff
@@ -7,22 +7,24 @@ const { PassThrough } = require('readable-stream')
...
+const util = require('util')
+const globPromise = util.promisify(glob)
 
 const dirList = require('./lib/dirList')
 
-function fastifyStatic (fastify, opts, next) {
+async function fastifyStatic (fastify, opts) {
   const error = checkRootPathForErrors(fastify, opts.root)
-  if (error !== undefined) return next(error)
+  if (error !== undefined) throw error
```
Comment (simoneb) [inline index.js:17]: minor. this pattern of returning an error was necessary in the callback based version but it's no longer necessary now. it could be changed so that the `checkRootPathForErrors` function immediately throws an error, meaning that if that function is invoked and no error is thrown, then the plugin registration can continue and there would be no need, here, to test the return value of that function. in any case, I appreciate this would require more changes so it's also ok if you leave it as it is.

File: index.js [hunk 13]
```diff
@@ -152,7 +164,7 @@ function fastifyStatic (fastify, opts, next) {
   if (opts.serve !== false) {
     if (opts.wildcard === undefined || opts.wildcard === true) {
       fastify.get(prefix + '*', routeOpts, function (req, reply) {
-        pumpSendToReply(req, reply, '/' + req.params['*'])
+        pumpSendToReply(req, reply, '/' + req.params['*'], sendOptions.root)
```
Comment (simoneb) [inline index.js:167]: why this change?

**Review by simoneb — APPROVED** [review]
This LGTM overall. I think you could improve it in a few places where you apply conditional logic based on whether the root option is a string or an array. In most cases you could treat both cases the same by turning the string into a singleton array and always assuming you're dealing with arrays from then on, thereby reducing by a great extent the conditional logic that you have in multiple places where you're checking whether it's a string or an array before deciding what to do with it.

File: index.js (hunk 13 above)
Comment (nigelhanlon) [reply index.js:167]: I think this was an early change I made that can be reverted. No valid reason I can think of to change it right now.

File: example/server-multi-dir-list.js (hunk 10 above)
Comment (nigelhanlon) [reply example/server-multi-dir-list.js]: I think we should keep it as a single file as it makes for an easy all in one reference. I have formatted the handlebar template in this example and `server-dir-list.js` to make them easier to read.

File: example/server-multi-dir-list.js (hunk 11 above)
Comment (nigelhanlon) [reply example/server-multi-dir-list.js]: Added comments to all examples 👍

File: index.js (hunk 12 above)
Comment (nigelhanlon) [reply index.js]: `checkRootPathForErrors` now immediately throws on any error with the root path option.

**Review by mcollina — APPROVED** [review]
lgtm

**Review by salmanm — APPROVED** [review]
